selfdrive/athena/athenad.py

```diff
@@ -100,35 +103,53 @@ def jsonrpc_handler(end_event):
 
 
 def upload_handler(end_event):
+  tid = threading.get_ident()
+
   while not end_event.is_set():
+    cur_upload_items[tid] = None
+
     try:
-      item = upload_queue.get(timeout=1)
-      if item.id in cancelled_uploads:
-        cancelled_uploads.remove(item.id)
+      cur_upload_items[tid] = upload_queue.get(timeout=1)._replace(current=True)
+      if cur_upload_items[tid].id in cancelled_uploads:
+        cancelled_uploads.remove(cur_upload_items[tid].id)
         continue
 
       try:
-        _do_upload(item)
-      except (requests.exceptions.Timeout, requests.exceptions.ConnectionError, requests.exceptions.SSLError) as e:
-        cloudlog.warning(f"athena.upload_handler.retry {e} {item}")
+        def cb(sz, cur):
+          cur_upload_items[tid] = cur_upload_items[tid]._replace(progress=cur / sz if sz else 1)
```

Comment:
maybe it is time to make `UploadItem` a class, put this callback method on it, and change all the `_replace()` calls to be setting props?

Reply:
I kinda like the immutability of the UploadItem named tuple actually. I'm still not a fan of modifying items while in the queue, that feels like an anti-pattern.

Reply:
yeah immutable objects are good for multi-threading, it just seems hard to follow how some of the pieces tie together, maybe the problem is mixing the thread state and the pending queue items